## Instructions 
## Create a Python script to visualize the weather of 500+ cities across the world of varying distance from the equator.
# Your objective is to build a series of scatter plots to showcase the following relationships:

#Temperature (F) vs. Latitude
#Humidity (%) vs. Latitude
#Cloudiness (%) vs. Latitude
#Wind Speed (mph) vs. Latitude
#Your final notebook must:

#Randomly select at least 500 unique (non-repeat) cities based on latitude and longitude.
#Perform a weather check on each of the cities using a series of successive API calls.
#Include a print log of each city as it's being processed with the city number, city name, and requested URL.
#Save both a CSV of all data retrieved and png images for each scatter plot.

In [ ]:
#1 Generate Cities List
#2 Perform API Calls
#3 Display the City Data Frame

#Banking Deserts example 

In [2]:
# Dependencies
import json
import requests
from citipy import citipy
from random import uniform
import pandas as pd
import matplotlib.pyplot as plt



#Import api_key 
from config import api_key

In [3]:
# Generate Cities List
cities = []
def newpoint():
    return uniform(-180, 180), uniform(-90, 90)

r_coordinates = (newpoint() for x in range(10))
for coordinate_pair in r_coordinates:
    #print (city)
    cities.append(coordinate_pair)
    

In [4]:
cities

[(-132.07149760283338, -54.56446126526445),
 (108.94917354810678, 85.70590992208255),
 (50.52244120063855, -7.6665839830436795),
 (39.906301872253636, 74.16326772473204),
 (-138.1195354836376, -40.26840891884196),
 (18.543846369751293, 65.82868422788363),
 (42.358868086321024, -6.014925450162835),
 (-120.94953097519132, -83.15965943255087),
 (165.76750710198655, -67.40412223967932),
 (179.59774813386974, -27.678247313324427)]

In [5]:
labels= ["Lon", "Lat"]
cities_coordinates = pd.DataFrame(cities, columns=labels)
cities_coordinates

Lon        Lat
0 -132.071498 -54.564461
1  108.949174  85.705910
2   50.522441  -7.666584
3   39.906302  74.163268
4 -138.119535 -40.268409
5   18.543846  65.828684
6   42.358868  -6.014925
7 -120.949531 -83.159659
8  165.767507 -67.404122
9  179.597748 -27.678247

In [10]:
#Specify the URL : Lat and Lon as parameters 
url = "http://api.openweathermap.org/data/2.5/weather?lat={lat}&lon={lon}"
units = "imperial"
# build query URL
query_url = f"{url}appid={api_key}&units={units}&q="
query_url



'http://api.openweathermap.org/data/2.5/weather?lat={lat}&lon={lon}appid=3c4a5bf4e16f21b1cccc13f0224ac2da&units=imperial&q='

In [20]:
weather_response = requests.get(query_url)
weather_json = weather_response.json()
weather_json

{'cod': 401,
 'message': 'Invalid API key. Please see http://openweathermap.org/faq#error401 for more info.'}

# Error message above AND below : HTTP Error 401: 'Invalid API key

# ...why? API key does work ...(used it to re-do class examples. Also generated a new key and that one does not work either....erroneous syntax? API key does work in my other Working file ("Working-Wrapper")


# Error Below: 

# Does ow take coordinates in (x,y) format or do we have to write them as they are return in JSON  ie. "coord.lat", "coord.lon" ? 



## "cities" is in form (lon, lat) 
## "cities_coordinates" (DF) is in form column Lon, column Lat 

# Figure out how to input coordinates: 1) (lon, lat) aka "cities" OR 2) column Lon, column Lat aka "cities_coordinates"

In [18]:
#List for response results 
city_name = []
city_id = []
temperature= []
humidity= []
cloudiness= []
wind_speed= []
city_url = []

# Loop through each city, Perform API Calls, Append desired results 

#for loop for cities = (long ,lat)
# for city in cities:


#for loop for cities_coordinate = dataframe
for city in cities_coordinates:
    response = requests.get(query_url + city).json()
    print(response)
    city_name.append(response["name"])
    city_id.append(response["sys.id"])
    temperature.append(response["main.temp"])
    humidity.append(response["humidity.unit"])
    cloudiness.append(response["clouds.all"])
    wind_speed.append["wind.speed"]
    
    
# ["name", "sys.id", "main.temp", 
# AN#  "humidity.unit", "clouds.all", "wind.speed"]

{'cod': 401, 'message': 'Invalid API key. Please see http://openweathermap.org/faq#error401 for more info.'}


KeyError: 'name'

# Another Route 
# except error still present b/c need to print URL
# Prints a URL but it's a 401 URL

In [21]:
# set up params
params = {
    "api_key": api_key
}

units = "imperial"

# Counter
row_count = 0
for index, row in cities_coordinates.iterrows():
    params['Lon'] = row['Lat']
    base_url = "http://api.openweathermap.org/data/2.5/weather?lat={lat}&lon={lon}"
    #except, error with printing URL
    print("Now retrieving city # " + str(row_count))
    row_count += 1
    lat_location = requests.get(base_url, params=params)
    print(lat_location.url)
    
    lat_location= lat_location.json()
    

Now retrieving city # 0
http://api.openweathermap.org/data/2.5/weather?lat=%7Blat%7D&lon=%7Blon%7D&api_key=3c4a5bf4e16f21b1cccc13f0224ac2da&Lon=-54.56446126526445
Now retrieving city # 1
http://api.openweathermap.org/data/2.5/weather?lat=%7Blat%7D&lon=%7Blon%7D&api_key=3c4a5bf4e16f21b1cccc13f0224ac2da&Lon=85.70590992208255
Now retrieving city # 2
http://api.openweathermap.org/data/2.5/weather?lat=%7Blat%7D&lon=%7Blon%7D&api_key=3c4a5bf4e16f21b1cccc13f0224ac2da&Lon=-7.6665839830436795
Now retrieving city # 3
http://api.openweathermap.org/data/2.5/weather?lat=%7Blat%7D&lon=%7Blon%7D&api_key=3c4a5bf4e16f21b1cccc13f0224ac2da&Lon=74.16326772473204
Now retrieving city # 4
http://api.openweathermap.org/data/2.5/weather?lat=%7Blat%7D&lon=%7Blon%7D&api_key=3c4a5bf4e16f21b1cccc13f0224ac2da&Lon=-40.26840891884196
Now retrieving city # 5
http://api.openweathermap.org/data/2.5/weather?lat=%7Blat%7D&lon=%7Blon%7D&api_key=3c4a5bf4e16f21b1cccc13f0224ac2da&Lon=65.82868422788363
Now retrieving city # 6
